samigo/samigo-impl/src/java/org/sakaiproject/samigo/impl/SamigoAvailableNotificationServiceImpl.java

```diff
@@ -0,0 +1,369 @@
+
+        String sender = "Sender: \"" + getServiceName() + "\" <" + getSetupRequest() + ">";
+        additionalHeaders.add(sender);
+        additionalHeaders.add("Content-type: text/html; charset=UTF-8");
+    }
+
+    public void scheduleAssessmentAvailableNotification(String publishedId) {
+
+        // Remove any previously scheduled notification
+        removeScheduledAssessmentNotification(publishedId);
+
+        try {
+            PublishedAssessmentFacade publishedAssessment = publishedAssessmentService.getPublishedAssessment(publishedId);
+            List<ExtendedTime> extensionContainer = PersistenceService.getInstance().getExtendedTimeFacade().getEntriesForPub(publishedAssessment.getData());
+            Date startDate = publishedAssessment.getStartDate();
+            // Only schedule a new notification if start date is in the future to prevent duplicates and spam
+            if (Instant.now().isBefore(startDate.toInstant())) {    //for main
+                scheduledInvocationManager.createDelayedInvocation(startDate.toInstant(), "org.sakaiproject.samigo.api.SamigoAvailableNotificationService", publishedId);
+            }
+            for (ExtendedTime extension: extensionContainer){  //make separate delayedInvocations for people with exceptions.
+                if (Instant.now().isBefore(extension.getStartDate().toInstant())) {
+                    scheduledInvocationManager.createDelayedInvocation(extension.getStartDate().toInstant(), "org.sakaiproject.samigo.api.SamigoAvailableNotificationService", publishedId + ',' + extension.getId().toString());
+                }
+            }
+        } catch (IllegalArgumentException e) {
+            log.error("Unable to schedule assessment open notification for assessment " + publishedId);
+        }
+    }
+
+    public void removeScheduledAssessmentNotification(String publishedId) {
+        scheduledInvocationManager.deleteDelayedInvocation("org.sakaiproject.samigo.api.SamigoAvailableNotificationService", publishedId);  //remove main reminder
+        PublishedAssessmentFacade publishedAssessment = publishedAssessmentService.getPublishedAssessment(publishedId);
+        if (publishedAssessment != null){
+            List<ExtendedTime> extensionContainer = PersistenceService.getInstance().getExtendedTimeFacade().getEntriesForPub(publishedAssessment.getData());
+            for (ExtendedTime extension: extensionContainer){  //remove all the assessment's extension reminders.
+                scheduledInvocationManager.deleteDelayedInvocation("org.sakaiproject.samigo.api.SamigoAvailableNotificationService", publishedId + ',' + extension.getId().toString());
+            }
+        }
+    }
+
+    public void execute(String publishedId) {
+        log.debug("Samigo available assessment email notification job starting");
+
+        Session sakaiSession = sessionManager.getCurrentSession();
+        sakaiSession.setUserId("admin");
+        sakaiSession.setUserEid("admin");
```

Comment:
This is very, very bad practise. Please, don't ever do this. Rather than hijacking the session to be admin, use a named `SecurityAdvisor`, and make sure you pop it off the stack in a `finally` block. There are many examples of this in Sakai already.